### PR TITLE
goodFeaturesToTrack returns also corner value

### PR DESCRIPTION
**Pull request by amirtu**
PR to extra with sanity data update: opencv/opencv_extra#841

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by amirtu** [comment]
This PR supports [https://github.com/opencv/opencv/pull/17836](#17836 goodFeaturesToTrack returns also corner value)

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 1]
```diff
@@ -1985,20 +1985,22 @@ pixel neighborhood. See cornerEigenValsAndVecs .
...
 
 @sa  cornerMinEigenVal, cornerHarris, calcOpticalFlowPyrLK, estimateRigidTransform,
  */
 
 CV_EXPORTS_W void goodFeaturesToTrack( InputArray image, OutputArray corners,
                                      int maxCorners, double qualityLevel, double minDistance,
                                      InputArray mask = noArray(), int blockSize = 3,
-                                     bool useHarrisDetector = false, double k = 0.04 );
+                                     bool useHarrisDetector = false, double k = 0.04,
+                                     OutputArray corners_values = noArray());
```
Comment (amirtu) [inline modules/imgproc/include/opencv2/imgproc.hpp:1997]: UB if we write at the noArray() while someone is reading?

File: modules/core/src/arithm.cpp [hunk 2]
```diff
@@ -1222,11 +1222,11 @@ void cv::compare(InputArray _src1, InputArray _src2, OutputArray _dst, int op)
             compare(_src2, _src1, _dst, op);
             return;
         }
-        else if(is_src1_scalar == is_src2_scalar)
+        else if(is_src1_scalar == true && is_src2_scalar == true)
             CV_Error( CV_StsUnmatchedSizes,
                      "The operation is neither 'array op array' (where arrays have the same size and the same type), "
                      "nor 'array op scalar', nor 'scalar op array'" );
-        haveScalar = true;
+        haveScalar = is_src1_scalar || is_src2_scalar;
```
Comment (asmorkalov) [inline modules/core/src/arithm.cpp:1229]: Please revert the changes and raise bug with arithmetic concern.

File: modules/imgproc/perf/opencl/perf_gftt.cpp [hunk 3]
```diff
@@ -72,14 +72,16 @@ OCL_PERF_TEST_P(GoodFeaturesToTrackFixture, GoodFeaturesToTrack,
...
     img.copyTo(src);
 
     declare.in(src, WARMUP_READ).out(dst);
 
     OCL_TEST_CYCLE() cv::goodFeaturesToTrack(src, dst, maxCorners, qualityLevel,
-                                             minDistance, noArray(), 3, 3, harrisDetector, 0.04);
+                                             minDistance, noArray(), 3, 3, harrisDetector, 0.04, values);
 
     SANITY_CHECK(dst);
+    SANITY_CHECK(values, 2.f * FLT_EPSILON, ERROR_RELATIVE);
```
Comment (asmorkalov) [inline modules/imgproc/perf/opencl/perf_gftt.cpp:84]: I propose to use absolute error 1e-6. It's small enough relatively to values and should cover floating computation issue on different OpenCL devices.

File: modules/imgproc/test/ocl/test_gftt.cpp [hunk 4]
```diff
@@ -100,22 +101,27 @@ OCL_TEST_P(GoodFeaturesToTrack, Accuracy)
...
 
         int mistmatch = 0;
         for (size_t i = 0; i < pts.size(); ++i)
         {
-            Point2i a = upts[i], b = pts[i];
+            Point2f a = upts[i], b = pts[i];
 
-            bool eq = std::abs(a.x - b.x) < 1 && std::abs(a.y - b.y) < 1;
+            bool eq = std::abs(a.x - b.x) < 1 && std::abs(a.y - b.y) < 1 &&
+                    std::abs(values[i] - uvalues[i]) <= 3.f * FLT_EPSILON * std::max(values[i], uvalues[i]);
```
Comment (asmorkalov) [inline modules/imgproc/test/ocl/test_gftt.cpp:124]: New condition increases amount of mismatches. I propose to tune Eps a little to pass original check.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
@vpisarev Please review the PR. @amirtu Cherry-picked original PR to presume original author and fixed found issues. We decided to create new PR because patch to extra repository is required. The PR introduces function overload instead of extra parameter with default value, because of API changes in Python. Python wrapper for function with 2 OutputArray parameter returns tuple of results and it breaks existing code. Overloads presume old behavior and introduces Python function with alternative name that returns tuple.

**Comment by amirtu** [comment]
To be merged with https://github.com/opencv/opencv_extra/pull/841

File: modules/imgproc/test/test_goodfeaturetotrack.cpp [hunk 5]
```diff
@@ -130,9 +130,10 @@ test_cornerEigenValsVecs( const Mat& src, Mat& eigenv, int block_size,
 
             for( j = 0; j < src.cols; j++ )
             {
-                double a = dx2p[j], b = dxdyp[j], c = dy2p[j];
-                double d = sqrt( ( a - c )*( a - c ) + 4*b*b );
-                eigenvp[j] = (float)( 0.5*(a + c - d));
+                double a = dx2p[j]*0.5f, b = dxdyp[j], c = dy2p[j]*0.5f;
+                //double d = sqrt( ( a - c )*( a - c ) + 4*b*b );
+                //eigenvp[j] = (float)( 0.5*(a + c - d));
+                eigenvp[j] = (float)((a + c) - std::sqrt((a - c)*(a - c) + b*b));
```
Comment (asmorkalov) [inline modules/imgproc/test/test_goodfeaturetotrack.cpp:136]: Please remove dead code.

File: modules/imgproc/include/opencv2/imgproc.hpp (hunk 1 above)
Comment (amirtu) [reply modules/imgproc/include/opencv2/imgproc.hpp]: Added check if writing is actuallly requested (seems the object does not allow write access by default anyway).

File: modules/imgproc/test/test_goodfeaturetotrack.cpp (hunk 5 above)
Comment (amirtu) [reply modules/imgproc/test/test_goodfeaturetotrack.cpp:134]: Done.

File: modules/imgproc/test/ocl/test_gftt.cpp (hunk 4 above)
Comment (amirtu) [reply modules/imgproc/test/ocl/test_gftt.cpp:124]: OK, tests passes with 3.f * FLT_EPSILON relative error threshold.

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 6]
```diff
@@ -1999,6 +1999,25 @@ CV_EXPORTS_W void goodFeaturesToTrack( InputArray image, OutputArray corners,
...
+@param corners_values Optional vector of detected corners response values.
+ */
+CV_EXPORTS CV_WRAP_AS(goodFeaturesToTrackWithResponse) void goodFeaturesToTrack(
+        InputArray image, OutputArray corners,
+        int maxCorners, double qualityLevel, double minDistance,
+        InputArray mask, int blockSize,
+        bool useHarrisDetector, double k,
+        OutputArray corners_values);
+
+CV_EXPORTS CV_WRAP_AS(goodFeaturesToTrackWithResponse) void goodFeaturesToTrack(
```
[the remark this answers is not in the document]
Comment (amirtu) [reply modules/imgproc/include/opencv2/imgproc.hpp:2033]: Done (I have added doc line comments to the first of the two functions, as previously only one of the two goodFeaturesToTrack() versions was documented).

File: modules/imgproc/perf/opencl/perf_gftt.cpp (hunk 3 above)
Comment (amirtu) [reply modules/imgproc/perf/opencl/perf_gftt.cpp]: Switched the tests back to comparison by absolute error.

**Review by asmorkalov — APPROVED** [review]
Well done! :+1:

**Comment by asmorkalov** [comment]
@vpisarev Could you take a look. Please ignore CN CI builds, they does not handle paired builds with extra right now.

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 7]
```diff
@@ -1999,6 +1999,44 @@ CV_EXPORTS_W void goodFeaturesToTrack( InputArray image, OutputArray corners,
...
+or #cornerMinEigenVal.
+@param k Free parameter of the Harris detector.
+@param corners_values Output vector of response values of the detected corners.
+ */
+CV_EXPORTS CV_WRAP_AS(goodFeaturesToTrackWithResponse) void goodFeaturesToTrack(
+        InputArray image, OutputArray corners,
+        int maxCorners, double qualityLevel, double minDistance,
+        InputArray mask, int blockSize,
+        bool useHarrisDetector, double k,
+        OutputArray corners_values);
```
Comment (alalek) [inline modules/imgproc/include/opencv2/imgproc.hpp:2031]: Do not violate OpenCV arguments order: https://github.com/opencv/opencv/wiki/Coding_Style_Guide
or make it optional.

> input parameters, output parameters, flags and optional parameters.

- required input arrays
- required output arrays
- required input flags / scalar / non-array parameters
- optional input arrays (mask)
- optional output arrays
- optional input flags / scalar / non-array parameters

Comment (alalek) [inline modules/imgproc/include/opencv2/imgproc.hpp:2031]: > corners_values

"values" are too abstract. All parameters are "values".
`corners_response`

File: modules/imgproc/perf/perf_goodFeaturesToTrack.cpp [hunk 8]
```diff
@@ -31,14 +31,16 @@ PERF_TEST_P(Image_MaxCorners_QualityLevel_MinDistance_BlockSize_gradientSize_Use
         FAIL() << "Unable to load source image" << filename;
 
     std::vector<Point2f> corners;
+    std::vector<float> corners_values;
 
     double minDistance = 1;
-    TEST_CYCLE() goodFeaturesToTrack(image, corners, maxCorners, qualityLevel, minDistance, noArray(), blockSize, gradientSize, useHarrisDetector);
+    TEST_CYCLE() goodFeaturesToTrack(image, corners, maxCorners, qualityLevel, minDistance, noArray(), blockSize, gradientSize, useHarrisDetector, 0.04, corners_values);
```
Comment (alalek) [inline modules/imgproc/perf/perf_goodFeaturesToTrack.cpp:37]: This change drops testing of old function.

You should create new one instead.

File: modules/imgproc/src/featureselect.cpp [hunk 9]
```diff
@@ -251,13 +254,18 @@ static bool ocl_goodFeaturesToTrack( InputArray _image, OutputArray _corners,
             const Corner & c = corner_ptr[i];
 
             corners.push_back(Point2f((float)c.x, (float)c.y));
+            corners_values.push_back(eig.getMat(ACCESS_READ).at<float>(c.y, c.x));
```
Comment (alalek) [inline modules/imgproc/src/featureselect.cpp:257]: > .getMat(ACCESS_READ)

`.getMat()` / `.getUMat()` in a loop is not a good idea

---

Did you run perf tests locally?

File: modules/imgproc/test/test_goodfeaturetotrack.cpp [hunk 10]
```diff
@@ -91,9 +91,9 @@ test_cornerEigenValsVecs( const Mat& src, Mat& eigenv, int block_size,
     denom = denom * denom;
 
     if( _aperture_size < 0 )
-        denom *= 4;
+        denom *= 2.;
     if(type != ftype )
-        denom *= 255.;
+        denom *= 255. * 255.;
```
Comment (alalek) [inline modules/imgproc/test/test_goodfeaturetotrack.cpp:96]: why?

**Review by alalek — COMMENTED** [review]
Modification of both implementation and tests is not a good idea in general.

You should keep already existed test cases to ensure that there are no regressions in already existed functions.
For new functionality add new tests (or add new test parameters with keeping old checks for old test cases).

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 11]
```diff
@@ -1999,6 +1999,42 @@ CV_EXPORTS_W void goodFeaturesToTrack( InputArray image, OutputArray corners,
...
+pixel neighborhood. See cornerEigenValsAndVecs .
+@param useHarrisDetector Parameter indicating whether to use a Harris detector (see #cornerHarris)
+or #cornerMinEigenVal.
+@param k Free parameter of the Harris detector.
+ */
+CV_EXPORTS CV_WRAP_AS(goodFeaturesToTrackWithQuality) void goodFeaturesToTrack(
+        InputArray image, OutputArray corners,
+        int maxCorners, double qualityLevel, double minDistance,
+        InputArray mask, OutputArray cornersQuality, int blockSize = 3,
+        bool useHarrisDetector = false, double k = 0.04 );
```
Comment (asmorkalov) [inline modules/imgproc/include/opencv2/imgproc.hpp:2030]: I think that this overload is redundant. The same functionality can be achieved with the next overload. The previous overload with and without `gradientSize` was added for backward compatibility. New functions does not require it.

File: modules/imgproc/test/test_goodfeaturetotrack.cpp (hunk 10 above)
Comment (asmorkalov) [reply modules/imgproc/test/test_goodfeaturetotrack.cpp:96]: @amirtu Please comment changes in reference code.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/imgproc/include/opencv2/imgproc.hpp (hunk 11 above)
Comment (amirtu) [reply modules/imgproc/include/opencv2/imgproc.hpp]: I have missed this, thanks for notice.

File: modules/imgproc/test/test_goodfeaturetotrack.cpp (hunk 10 above)
Comment (amirtu) [reply modules/imgproc/test/test_goodfeaturetotrack.cpp]: @alalek @asmorkalov Sorry for delayed answer, it has taken me some time to understand the context. Here are approximations of partial derivatives being computed with Sobel operator, the scaling is needed to preserve the energy after this filter: power of two due to averaging and differentiation parts of the filter, block size is considered in advance due to windowing of the image for every point, CV_32C1 is (probably) normalized, CV_8U is not. First derivatives in both horizontal and vertical direction then should be scaled, as can be seen in cornerEigenValsVecs() (/imgproc/.../corner.cpp).

**Comment by amirtu** [comment]
new PR for merge in 3.4: https://github.com/opencv/opencv/pull/19511

**Review by asmorkalov — APPROVED** [review]
Looks good for me. :+1:  Please squash your own commits and presume commits of original author.

File: modules/features2d/src/gftt.cpp [hunk 12]
```diff
@@ -106,14 +107,15 @@ class GFTTDetector_Impl CV_FINAL : public GFTTDetector
...
         }
 
         keypoints.resize(corners.size());
         std::vector<Point2f>::const_iterator corner_it = corners.begin();
         std::vector<KeyPoint>::iterator keypoint_it = keypoints.begin();
-        for( ; corner_it != corners.end() && keypoint_it != keypoints.end(); ++corner_it, ++keypoint_it )
-            *keypoint_it = KeyPoint( *corner_it, (float)blockSize );
+        std::vector<float>::iterator quality_it = cornersQuality.begin();
+        for( ; corner_it != corners.end() && keypoint_it != keypoints.end() && quality_it != cornersQuality.end(); ++corner_it, ++keypoint_it, ++quality_it )
+            *keypoint_it = KeyPoint( *corner_it, (float)blockSize, -1, *quality_it );
```
Comment (alalek) [inline modules/features2d/src/gftt.cpp:118]: This part of code looks strange (it was strange before this patch too).

What should we do with "left" uninitialized items when `keypoint_it != keypoints.end() && quality_it != cornersQuality.end()` is triggered instead of `corner_it != corners.end()`?

Perhaps it makes sense to put after this loop consistency check:

```
CV_Assert(corner_it == corners.end());
```

File: modules/imgproc/test/test_goodfeaturetotrack.cpp [hunk 13]
```diff
@@ -488,6 +502,19 @@ int CV_GoodFeatureToTTest::validate_test_results( int test_case_idx )
         ts->set_failed_test_info(cvtest::TS::OK);
     }
 
+    e = cv::norm(cornersQuality, RefcornersQuality, NORM_RELATIVE | NORM_INF);
+
+    if (e > eps)
+    {
+        ts->printf(cvtest::TS::CONSOLE, "actual error: %g, expected: %g", e, eps);
+        ts->set_failed_test_info(cvtest::TS::FAIL_BAD_ACCURACY);
```
Comment (alalek) [inline modules/imgproc/test/test_goodfeaturetotrack.cpp:510]: Prefer to use GTest's `EXPECT_LE(e, eps);` instead of legacy "ts->printf()".
It would should failed line at least.

**Review by alalek — APPROVED** [review]
Thank you! Looks good to me :+1:

File: modules/features2d/src/gftt.cpp (hunk 12 above)
Comment (amirtu) [reply modules/features2d/src/gftt.cpp]: Noted with thanks

File: modules/imgproc/test/test_goodfeaturetotrack.cpp (hunk 13 above)
Comment (amirtu) [reply modules/imgproc/test/test_goodfeaturetotrack.cpp]: Fixed, but unfortunately the test itself does not extend GTest so semantics of EXPECT_LE looks oddly.

File: modules/features2d/src/gftt.cpp [hunk 14]
```diff
@@ -106,14 +107,14 @@ class GFTTDetector_Impl CV_FINAL : public GFTTDetector
...
         }
 
-        keypoints.resize(corners.size());
-        std::vector<Point2f>::const_iterator corner_it = corners.begin();
-        std::vector<KeyPoint>::iterator keypoint_it = keypoints.begin();
-        for( ; corner_it != corners.end() && keypoint_it != keypoints.end(); ++corner_it, ++keypoint_it )
-            *keypoint_it = KeyPoint( *corner_it, (float)blockSize );
+        CV_Assert(corners.size() == cornersQuality.size());
+
+        keypoints.clear();
```
Comment (alalek) [inline modules/features2d/src/gftt.cpp:115]: `.reserve()` is necessary to avoid multiple re-allocations.

Comment (amirtu) [reply modules/features2d/src/gftt.cpp]: Fixed.
